### DISCUSSION TITLE
Indexed type access with Zod

### DISCUSSION THREAD

**Discussion by mia1024** [colinhacks/zod#1966] [Q&A] [answered]
Consider the following typescript code, which represents some data entry inside a database and the update command that is sent over the network (untrusted) 
```ts
interface Data {
    id: string;
    prop1: number;
    prop2: string;
    prop3: boolean;
}

interface DataUpdate<T extends keyof Data> {
    target: string; // the id
    prop: T;
    val: Data[T];
}
```

In this situation, it would be ideal to be able to use `zod` to validate the data, since it's transmitted over the network. However, it does not seem like there is a way currently to do an indexed type access with `zod`. While we can currently do the rest of the objects using the following code, we can't really do the `val` property with zod. 

```ts
const Data = z.object({
    id: z.string(),
    prop1: z.number(),
    prop2: z.string(),
    prop3: z.boolean(),
});

const DataUpdate = z.object({
    id: z.string(),
    prop: Data.keyof(),
});
```

Is this a new feature that can be added? Or am I missing something here?

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const Data = z.object( {
    id: z.string(),
    prop1: z.number(),
    prop2: z.string(),
    prop3: z.boolean(),
} )

const DataUpdate = z.object( {
    id: Data.shape.id,
    prop: Data.keyof(),
    val: z.unknown(),
} ).refine(
    data => Data.shape[ data.prop ].safeParse( data.val ).success,
    { message: 'Invalid value', path: [ 'val' ] }
)

console.log( DataUpdate.parse( { id: 'foo', prop: 'prop1', val: 123 } ) )
// { id: 'foo', prop: 'prop1', val: 123 }

const result = DataUpdate.safeParse( { id: 'foo', prop: 'prop1', val: 'foo' } )
!result.success && console.log( result.error.issues )
// [ { code: 'custom', message: 'Invalid value', path: [ 'val' ] } ]
```

**Comment by kwshi** [comment]
Another alternative-- after all, `DataUpdate` is a parametrized/generic/polymorphic type, which means that at the type level it's a _function_ that takes in a type `T` and returns a "version" of the monomorphic type for each given `T`. So here's how I would've thought to define it:

```typescript
const DataUpdate = (T: keyof z.TypeOf<typeof Data>) => z.object({
  id: Data.shape.id,
  prop: z.literal(T),
  val: Data.shape[T],
});
```

This is the most literal translation from static type to runtime type schema, with a tradeoff to the `refine` approach:
- pro: it preserves the type constraint _statically_ (the `refine` approach does validate things correctly at runtime, but the object returned still has non-constrained type `{prop: 'prop1' | 'prop2' | 'prop3', val: unknown}, I think).
- con: it doesn't give you automatic type inference for the generic argument; instead you have to explicitly specify `T` to construct a monomorphic instance of `DataUpdate` for each key you care about.

**Reply by JacobWeisenburger** [reply]
I really like your idea. Also the way you described everything is amazing. Great Work!

**Comment by JacobWeisenburger** [comment, marked as answer]
Another alternative [`z.discriminatedUnion`](https://github.com/colinhacks/zod#discriminated-unions)

```ts
const dataUpdateBase = z.object( { id: z.string() } )

const dataUpdate = z.discriminatedUnion( 'prop', [
    dataUpdateBase.extend( { prop: z.literal( 'prop1' ), val: z.number() } ),
    dataUpdateBase.extend( { prop: z.literal( 'prop2' ), val: z.string() } ),
    dataUpdateBase.extend( { prop: z.literal( 'prop3' ), val: z.boolean() } ),
] )

type DataUpdate = z.infer<typeof dataUpdate>
// type DataUpdate = {
//     id: string
//     prop: 'prop1'
//     val: number
// } | {
//     id: string
//     prop: 'prop2'
//     val: string
// } | {
//     id: string
//     prop: 'prop3'
//     val: boolean
// }

console.log( dataUpdate.parse( { id: 'foo', prop: 'prop1', val: 123 } ) )
// { id: 'foo', prop: 'prop1', val: 123 }

const result = dataUpdate.safeParse( { id: 'foo', prop: 'prop1', val: 'foo' } )
!result.success && console.log( result.error.issues )
// [
//     {
//       code: 'invalid_type',
//       expected: 'number',
//       received: 'string',
//       path: [ 'val' ],
//       message: 'Expected number, received string'
//     }
// ]
```
